nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -0,0 +1,95 @@
+package com.uber.nullaway;
+
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.AnnotatedTypeTree;
+import com.sun.source.tree.AnnotationTree;
+import com.sun.source.tree.ParameterizedTypeTree;
+import com.sun.source.tree.Tree;
+import com.sun.tools.javac.code.Attribute;
+import com.sun.tools.javac.code.Type;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/** Methods for performing checks related to generic types and nullability. */
+public class GenericsChecks {
+
+  private GenericsChecks() {
+    // just utility methods
+  }
+
+  /**
+   * Checks that for an instantiated generic type, {@code @Nullable} types are only used for type
+   * variables that have a {@code @Nullable} upper bound.
+   *
+   * @param tree the tree representing the instantiated type
+   * @param state visitor state
+   * @param analysis the analysis object
+   * @param config the analysis config
+   */
+  public static void checkInstantiationForParameterizedTypedTree(
+      ParameterizedTypeTree tree, VisitorState state, NullAway analysis, Config config) {
+    if (!config.isJSpecifyMode()) {
+      return;
+    }
+    List<? extends Tree> typeArguments = tree.getTypeArguments();
+    if (typeArguments.size() == 0) {
+      return;
+    }
+    Map<Integer, Tree> nullableTypeArguments = new HashMap<>();
+    for (int i = 0; i < typeArguments.size(); i++) {
+      Tree curTypeArg = typeArguments.get(i);
+      if (curTypeArg instanceof AnnotatedTypeTree) {
+        AnnotatedTypeTree annotatedType = (AnnotatedTypeTree) curTypeArg;
+        for (AnnotationTree annotation : annotatedType.getAnnotations()) {
+          Type annotationType = ASTHelpers.getType(annotation);
+          if (annotationType != null
+              && Nullness.isNullableAnnotation(annotationType.toString(), config)) {
+            nullableTypeArguments.put(i, curTypeArg);
+            break;
+          }
+        }
+      }
+    }
+    // base type that is being instantiated
+    Type baseType = ASTHelpers.getType(tree);
+    if (baseType == null) {
+      return;
+    }
+    com.sun.tools.javac.util.List<Type> baseTypeArgs = baseType.tsym.type.getTypeArguments();
+    for (int i = 0; i < baseTypeArgs.size(); i++) {
+      if (nullableTypeArguments.containsKey(i)) {
+
+        Type upperBound = baseTypeArgs.get(i).getUpperBound();
+        com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrors =
+            upperBound.getAnnotationMirrors();
+        boolean hasNullableAnnotation =
+            Nullness.hasNullableAnnotation(annotationMirrors.stream(), config);
+        // if base type argument does not have @Nullable annotation then the instantiation is
+        // invalid
+        if (!hasNullableAnnotation) {
+          invalidInstantiationError(nullableTypeArguments.get(i), state, analysis);
+        }
+      }
+    }
+    // recursive check for nested parameterized types
```

Comment:
So, is this necessary because `matchParameterizedType(...)` only matches the top-level `ParameterizedTypeTree`? Did we check this was the case? If so, it might be worth documenting in the comment.

Reply:
Yes. It only checks for the top level `paramterizedTypedTree`. I will add comments.

Reply:
Actually @lazaroclapp is right!  @NikitaAware the thing that changed here is that `NullAway` now implements `com.google.errorprone.bugpatterns.BugChecker.ParameterizedTypeTreeMatcher`.  This means that Error Prone will call `NullAway.matchParameterizedType()` for _every_ `ParameterizedTypeTree` including nested ones.  So, we can just delete this whole loop.  I tested locally and all our regression tests pass.

Before, we were trying to find all places where a `ParameterizedTypeTree` might occur, and then we had to manually recurse.  But the current approach is much simpler; I just didn't realize it was possible before.

Reply:
ok got it. Thanks.